### DISCUSSION TITLE
Controlling and understanding parallel requests

### DISCUSSION THREAD

**Discussion by bodograumann** [br3ndonland/inboard#87] [Q&A] [answered]
Due to some issues with requests that take a long time, we wondered about how many requests are actually run in parallel / concurrently. We don't necessarily want to have a large limit, but we want to be able to control the maximal number of requests processed in parallel.
We are using fastapi and had 3 guincorn workers defined, but there were 7 requests processed in parallel. My guess is that this is 8 cores - 1, but I couldn't find any place where this is configured explicitely. During my research only more questions popped up.

FastAPI uses Starlette, which does parallelization with a thread-pool using AnyIO. So why does it make sense to still have gunicorn with multiple workers on top? (In this case we define our controller method without asyncio, so no concurrency is used.)

The default AnyIO threadpool has size 40, so that is not the limiting factor.

Uvicorn also has some options for `workers` and `limit_concurrency`, but I don't understand what these would do in the complex context...

Could anybody enlighten me?

**Comment by bodograumann** [comment]
I did some [testing](https://github.com/bodograumann/fastapi-parallel-testbed) and found the following.

Most likely the observation of a maximum of 7 parallel requests handled was not correct and all requests received by the service where handled in parallel.

There is a vital difference between using `def` and `async def` for endpoints in FastAPI.

When using `def`, starlette (which FastAPI is based upon) uses an AnyIO threadpool to process multiple requests in parallel. The limit of parallel requests is the size of the threadpool (so 40 by default, which is difficult to change). These execution paths share the same FastAPI app instance, python modules and any global variables. The GIL prevents most serious problems, but you might still have to use locks.

When using `async def` there is only one thread per gunicorn worker with a single FastAPI instance. In this case multiple requests might be handled concurrently at the places where `await` is used. Within the non-concurrent parts of the code you don't have to worry about synchronization, but you might still want to be careful when using global state.

I didn't find a built-in way in FastAPI to limit the concurrency per worker, but a standard `asyncio.Semaphore` could probably be used. One open question here would be how gunicorn assigns requests to workers. When a worker is blocked, successive requests should be handled by the other workers instead. At the point the semaphore blocks this is probably too late.

Sources:
- https://stackoverflow.com/questions/70446584/how-does-fastapi-uvicorn-parallelize-requests
- https://stackoverflow.com/questions/71516140/fastapi-runs-api-calls-in-serial-instead-of-parallel-fashion
- https://stackoverflow.com/questions/48483348/how-to-limit-concurrency-with-python-asyncio
- https://anyio.readthedocs.io/en/stable/synchronization.html#capacity-limiters

**Comment by br3ndonland** [comment, marked as answer]
Thank you for sharing some insights on this. Concurrency can be difficult to understand and control.

> There is a vital difference between using `def` and `async def` for endpoints in FastAPI.

Yes, definitely. I think it's summarized well in the [post you shared](https://stackoverflow.com/questions/71516140/fastapi-runs-api-calls-in-serial-instead-of-parallel-fashion):

> **Note** that `async` does not mean _parallel_, but _concurrently_. Asynchronous code with [`async` and `await` is many times summarised as using coroutines](https://fastapi.tiangolo.com/async/#coroutines). **Coroutines** are collaborative (or [cooperatively multitasked](https://en.wikipedia.org/wiki/Cooperative_multitasking)), meaning that "at any given time, a program with coroutines is running **only** one of its coroutines, and this running coroutine suspends its execution only when it explicitly requests to be suspended" (see [here](https://stackoverflow.com/questions/553704/what-is-a-coroutine) and [here](https://stackoverflow.com/questions/1934715/difference-between-a-coroutine-and-a-thread) for more info on coroutines)...
>
> If, however, a blocking I/O-bound or CPU-bound operation was directly executed/called inside an `async def` function/endpoint, it would **block the main thread** (and hence, the `event loop`). Hence, a blocking operation such as `time.sleep()` in an `async def` endpoint would block the entire server (as in the code example provided in your question). Thus, if your endpoint is not going to make any async calls, you could declare it with just def instead, which would be run in an external threadpool that would then be awaited...

This means that `def` or `async def` should be chosen based on _what the function does_.

- When a function contains entirely sync logic, using `def` instead of `async def` will allow Starlette to run the function in its threadpool.
- When a function contains async logic with `await` statements, `async def` must be used. But what if an `async def` function has multiple `await`s? If not handled correctly, the function can actually return as slowly as a standard synchronous function. One strategy for handling multiple `await`s is using an [AnyIO task group](https://anyio.readthedocs.io/en/stable/tasks.html). This is not explained well in the FastAPI or Starlette docs. Dagger's concurrency guide may be helpful (https://github.com/dagger/dagger/issues/4766).
- When a function mixes sync and async logic, a threadpool executor or `starlette.concurrency.run_in_threadpool` may be helpful (https://github.com/tiangolo/fastapi/discussions/7623#discussioncomment-5142888).

> Due to some issues with requests that take a long time, we wondered about how many requests are actually run in parallel / concurrently. We don't necessarily want to have a large limit, but we want to be able to control the maximal number of requests processed in parallel.

There is some interest in customizing the size of the Starlette threadpool (https://github.com/encode/starlette/issues/1724).

> I didn't find a built-in way in FastAPI to limit the concurrency per worker

Each Gunicorn worker runs a separate FastAPI app instance. If using a Uvicorn worker with Gunicorn, concurrency limits can be specified through Uvicorn. Uvicorn has a `limit_concurrency` config setting that may help, but Uvicorn settings cannot be customized by default when using workers because [`uvicorn.workers.UvicornWorker` has hard-coded `CONFIG_KWARGS`](https://github.com/encode/uvicorn/blob/d3cb170262b9da62db9b5a573b6abcee4e5f3fcc/uvicorn/workers.py#L14C7-L20).

The solution recommended in the [Uvicorn docs](https://www.uvicorn.org/deployment/#gunicorn) is to subclass `uvicorn.workers.UvicornWorker` and modify the `CONFIG_KWARGS`. It might look like this:

```py
from uvicorn.workers import UvicornWorker


class MyUvicornWorker(UvicornWorker):
    CONFIG_KWARGS = {"loop": "auto", "http": "auto", "limit_concurrency": 30}

```

Let's say this subclass is in `my_package.MyUvicornWorker`. [When running inboard, specify the environment variable `WORKER_CLASS="my_package.MyUvicornWorker"`](https://inboard.bws.bio/environment#gunicorn). This will be passed to the [Gunicorn `worker_class` setting](https://docs.gunicorn.org/en/latest/settings.html#worker-class). Gunicorn will run the custom worker and apply its config settings.
